typescript-test-samples/schema-and-contract-testing/README.md

```diff
@@ -0,0 +1,109 @@
+
+With a small number of producers and consumers, it might be easy to collaborate to avoid breaking changes. But as time goes on your event-driven architectures will grow and keeping on top of producer and consumer relationships can be hard. Hence it is important to automate schema validation to prevent breaking changes. 
+
+### System Under Test (SUT)
+
+This test is integration service agnostic, as it runs against the schema published to a schema registry. Amazon EventBridge Schema Registry is one example of such a registry. It provides schemas for all events that are generated by AWS services. When you turn on schema discovery it will automatically detect and provide schemas for your custom events (OpenAPI and JSON Schemas). You can also use your own Git repository as a registry for storing schemas. In our test we will use a local “schemas” folder as a registry for simplicity. 
+
+Schema testing checks for changes that are not backward-compatible, such as removing elements, changing a field type or adding new mandatory elements. Adding optional elements is considered backward-compatible.
+
+The diagram below shows an example of running the schema validation as part of a producer’s CI/CD pipeline. A sample unit test from the "Build" step is provided in this pattern - ![Schema Diff Unit Test](tests/schema-testing.test.ts).
+
+
+![System Under Test (SUT)](img/schema_testing.png) 
+
+### Limitations
+
+First, schema inherently has ambiguity as does not capture business constraints. It does not tell if particular combination of fields is invalid, or make explicit dependencies between fields. As an example, when you create a AWS Lambda function, the runtime property is mandatory for zip packaging formats but not required for container image packaging. The schema for create function request payload will not capture this. There is always the possibility of business logic breaking on consumer side due to assumptions made about payload even though there are no breaking schema changes.
+
+Second, unless consumer uses SDK/code bindings generated specifically for the version of schema in use, there is a likelihood of drift between code and schema. There is no way to guarantee the business logic adhere to the schema. Schema validation results in a false sense of security in such cases.
+
+Both of these limitations can be addressed with [Contract testing](#contract-testing).
+
+## Contract Testing
+
+### System Under Test (SUT)
+
+This test is integration service agnostic as it uses sample events generated from schema published to a schema registry. The diagram below shows one way of implementing contract testing as part of producer’s CI/CD pipeline.
+
+The diagram below shows an example of running the schema validation as part of a producer’s CI/CD pipeline. A sample unit test from the "Build" step is provided in this pattern - ![Contract Unit Test](tests/contract-testing.test.ts).
```

Comment:
Link here is broken, same as previous section.

Reply:
fixed